Email validation done

lkbhitesh07 opened the pull request:
<!-- Please remove any unused sections.

Note that anything written between these symbols will not appear in the actual, published PR. They serve as instructions for filling out this template. You may want to use the 'preview' tab above this textbox to verify formatting before submitting.
-->

## Summary
### Description of the change(s) you made
<!-- Briefly summarize your changes in 1-2 sentences here. -->
The problem we were facing regarding the validation of email, I just simply added the validation in which it will be checking if there is other then the actual email it will raise the error.


### Screenshots (if applicable)
<!-- If not applicable, please delete this section -->
![Screenshot from 2021-03-17 02-28-41](https://user-images.githubusercontent.com/44300735/111379180-d597ab00-86c8-11eb-9e9e-6bee37221799.png)

___
## Reviewer guidance
### How can a reviewer test these changes?
<!-- If not applicable, please delete this section -->
1. Go to channel edit page
2. Select 'share channel' from top right drop down menu
3. Type in email for an existing user but surrounded by <>
4. See the error.


## References
<!--
Additional, helpful things to add in this section:
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
 * references to relevant issues or Notion projects
-->

## Comments
<!-- Additional comments may be added here -->

----

### Contributor's Checklist
<!-- After saving the PR, come through to tick off completed checklist items. Delete any sections that are not applicable to your PR -->

PR process:

- [ ] If this is an important user-facing change, PR or related issue the `CHANGELOG` label been added to this PR. Note: items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] If this includes an internal dependency change, a link to the diff is provided
- [ ] The `docs` label has been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] If any Python requirements have changed, the updated `requirements.txt` files also included in this PR
- [ ] Opportunities for using Google Analytics here are noted
- [ ] Migrations are [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/)

Studio-specifc:

- [ ] All user-facing strings are translated properly
- [ ] The `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)
- [ ] All UI components are LTR and RTL compliant
- [ ] Views are organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)
- [ ] Users' storage used is recalculated properly on any changes to main tree files
- [ ] If there new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text), it has been noted.


Testing:

- [ ] Code is clean and well-commented
- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Any new interactions have been added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)
- [ ] Critical and brittle code paths are covered by unit tests
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependency files were updated if necessary (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

rtibbles commented on contentcuration/contentcuration/frontend/shared/views/channel/ChannelSharing.vue:
```diff
@@ -83,6 +83,7 @@
         shareMode: SharingPermissions.EDIT,
         error: null,
         sharing: false,
+        reg: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,24}))$/
```
It is reasonably possible that this regex will prevent some valid email addresses being included.

I would suggest taking a more restrictive approach, in order to solve the example posted in the issue.

The principal issue was email addresses wrapped in `<a@a.com>` as `>` is not a valid character for the domain part of the email address, it is probably sufficient to add exclusion of that in the latter part of the email address, after `@`, to the regex.

Also, if you could add a test case for the `<a@a.com>` style email to the test suite: https://github.com/learningequality/studio/blob/hotfixes/contentcuration/contentcuration/frontend/shared/views/channel/__tests__/channelSharing.spec.js#L75 that would help to give some assurance that this fix resolves this issue.

lkbhitesh07 replied:
@rtibbles, I'm looking for one more decent approach to solve this issue, I'll update you once done if not I'll update you with the changes you suggested.
Thanks

rtibbles replied:
The simple approach is sufficient - the general consensus that I can see is that email validation is a hard problem, so we shouldn't try to find a one-size fix all solution, but rather prevent common user mistakes (which we will discover as people make them!)

lkbhitesh07 replied:
@rtibbles, You were right this approach is restricting some valid email addresses, I'll just add these `< >` and will update in the future.

lkbhitesh07 replied:
@rtibbles, Sorry to ping you again but I have a small doubt, I was playing around with the regex `else if (!/.+@.+\..+/.test(this.email))` so I removed `!` to check if it'll give an error but I noticed that it's not giving an error which means it was giving same results both the time(we remove or we don't `!`). Will you please confirm it?
Also, I usually get a bit confuse in the regex part will you please tell me where should I add `>` to `/.+@.+\..+/`
Thanks

rtibbles reviewed (changes requested):
Can do a simpler solution initially, with a specifically targeted test case.

lkbhitesh07 commented:
Done @rtibbles, Please take a look. Also, I have tested it on my local machine it's working perfectly fine.
Thanks

rtibbles commented:
Hi @lkbhitesh07 - you've not added a test case for the specific issue you are trying to fix as I requested, and your removal of `!` from the regex is meaning that the current test cases are now failing. It will now flag valid email addresses as invalid.

lkbhitesh07 commented:
Hello @rtibbles, Sorry for not leaving an explanation on why I removed `!` and added `>` in the end, and yes you are right it's quite obvious that this will only check for the type of this email` <a@a.com>` and will fail for other wrong addresses.
So how we'll be able to restrict `>` in our regex because `!/.+@.+\..+/.test(this.email)` this is only checking that if the characters in our email contain characters in our regex but it's not validating that our email should only follow this pattern, can I get some insights on how should I tackle it?
Also regarding adding the test case, did you mean to replace `not a real email` with `<a@a.com>` in here `wrapper.setData({ email: 'not a real email', loading: false });`? I got bit confused here.
Thanks

rtibbles commented:
Closing as stale.
